### PR TITLE
userHasVisibility & createLibraryFromTag

### PR DESCRIPTION
**Pull request by ahsu1230**
commander functions & tests
for (1) check to see if user has visibility to a library
for (2) create a library from a tag (follows the same structure as copy/move keeps to a library)
if a library of same tag name exists, any URIs that already exist in the library will have an error
these new keeps in the destination library will also retain the tag

@andrewconner 

controller API still in the works...

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala [hunk 1]
```diff
@@ -180,6 +221,19 @@ class LibraryCommander @Inject() (
...
 
+  def userHasVisibility(userId: Id[User], libraryId: Id[Library]): Boolean = {
+    db.readOnlyMaster { implicit s =>
+      libraryRepo.get(libraryId).visibility match {
+        case LibraryVisibility.PUBLISHED =>
+          true
+        case LibraryVisibility.DISCOVERABLE =>
+          libraryMembershipRepo.getWithLibraryIdAndUserId(libraryId, userId).nonEmpty
+        case LibraryVisibility.SECRET =>
+          false
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala]: What about owners of libraries? Or people that are followers of secret libraries?

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala [hunk 2]
```diff
@@ -160,16 +161,56 @@ class LibraryCommander @Inject() (
...
+        }
+        (lib, badKeeps.toSeq)
     }
   }
 
-  def getLibraryByUserAndSlug(ownerId: Id[User], slug: LibrarySlug): Option[Library] = {
-    db.readOnlyMaster { implicit s =>
-      libraryRepo.getBySlugAndUserId(userId = ownerId, slug = slug)
-    }
+  def getLibraryById(id: Id[Library]): Library = db.readOnlyMaster { implicit s =>
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala]: It may be better not to have this. It's not really doing anything besides opening a db connection and calling a method. Anywhere we have that needs the method call can just do it itself, and potentially re-use a db connection (instead of forcing a new one).

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala [hunk 3]
```diff
@@ -160,16 +161,56 @@ class LibraryCommander @Inject() (
     }
   }
 
-  def getLibraryById(id: Id[Library]): Library = {
-    db.readOnlyMaster { implicit s =>
-      libraryRepo.get(id)
+  def createLibraryFromCollection(userId: Id[User], tag: Collection): (Library, Seq[(Keep, LibraryError)]) = db.readWrite { implicit s =>
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala]: What if this is built on top of creating a library? So the user will pass in the name, slug, discoverability, etc, and that way they can do the transition however they like.

Also, we're on the way to dropping the `collection` table (I should have mentioned this), since tags will become just freeform text. This method may be better to just take a tag name, and call `collectionRepo.getByUserAndName` so we can migrate easily when it's time.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala (hunk 2 above)
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala]: Ok, for now it's just only being used by LibraryController.getLibraryById

File: kifi-backend/modules/shoebox/test/com/keepit/commanders/LibraryCommanderTest.scala [hunk 4]
```diff
@@ -345,6 +345,19 @@ class LibraryCommanderTest extends Specification with ShoeboxTestInjector {
       }
     }
 
+    "does user have visibility" in {
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/test/com/keepit/commanders/LibraryCommanderTest.scala:333]: Mind adding a test checking for secret libraries? Users should have access to their own (and following) secret libraries

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala (hunk 1 above)
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala]: oh right about owners. But people can follow a secret library?

**Comment by andrewconner** [comment]
Nice, you're moving quickly. I didn't realize you were ready for these two. Tag -> Library is basically an import mechanism so that current users can take existing tags and create new libraries (and if we have time: add to existing libraries). It'll re-use the "create library" dialog on the front-end, but will make a different API call optionally. An idea is to have the client make two API calls: 1) creates the library as usual, no change, 2) tells the backend to copy all the keeps with a tag "tagname" to the library.

This way, adding to an existing library is easy. What do you think?

**Comment by andrewconner** [comment]
For the 2nd one, user visibility, no need for a controller method. Just commander. It's for internal ACLs only. Clients won't make this request directly; rather, they'll request for what they need, and get an error if permission denied.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala (hunk 1 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala]: Yep, if invited.

"Published" = library is visible to all, and is discoverable to all (see the Google doc for explicit meaning of discoverable)
"Discoverable" = visible to only members, but discoverable to friends
"Secret" = visible to only members, not discoverable to anyone but members

**Comment by ahsu1230** [comment]
I actually like the calling the two APIs better. It would be easier to handle all the edge cases (like if a slug is already used, etc.) if the API is separated into the two instead of all one big chunk.

File: kifi-backend/modules/shoebox/test/com/keepit/commanders/LibraryCommanderTest.scala (hunk 4 above)
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/test/com/keepit/commanders/LibraryCommanderTest.scala:333]: yup will do

**Comment by andrewconner** [comment]
Cool let's do that then. So in the tag -> library API call, we take in a library ID and tag name.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala [hunk 5]
```diff
@@ -160,16 +161,41 @@ class LibraryCommander @Inject() (
...
+              val newKeep = keepRepo.save(Keep(
+                userId = ownerId,
+                libraryId = Some(libraryId),
+                title = k.title,
+                uriId = k.uriId,
+                urlId = k.urlId,
+                url = k.url,
+                isPrivate = k.isPrivate,
+                bookmarkPath = k.bookmarkPath,
+                source = KeepSource.site))
```
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala]: would source be different?

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala [hunk 6]
```diff
@@ -160,16 +161,41 @@ class LibraryCommander @Inject() (
...
+              keepToCollectionRepo.save(KeepToCollection(keepId = newKeep.id.get, collectionId = tag.id.get))
```
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala]: make sure new keeps copied to library retain tag?

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala (hunk 5 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala]: Oo, good question. Simple solution is make a new `KeepSource` for tag import.

This brings up another idea: let users pick if they want the keeps affected to be copied, or moved. If moved, just trigger moving keeps between libraries. If copied, do exactly what you're doing. What do you think?

Also, any way to re-use move/copy commander methods, or would that be a pain?

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala (hunk 6 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala]: For now, I think that's a good policy. Good thinking.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala (hunk 5 above)
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala]: I might be able to reuse that code. the one in copy/move right now checks a lot of permissions between the source/destination libraries

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala [hunk 7]
```diff
@@ -160,16 +161,29 @@ class LibraryCommander @Inject() (
...
+          Left(LibraryFail("tag not found"))
+        case Some(tag) =>
+          val keeps = keepToCollectionRepo.getByCollection(tag.id.get).map(k2c => keepRepo.get(k2c.keepId))
+          def saveKeep(k: Keep, s: RWSession): Unit = {
+            implicit val session = s
+            val newKeep = keepRepo.save(Keep(title = k.title, uriId = k.uriId, url = k.url, urlId = k.urlId, isPrivate = k.isPrivate,
+              userId = k.userId, source = KeepSource.tagImport, libraryId = Some(libraryId)))
+            keepToCollectionRepo.save(KeepToCollection(keepId = newKeep.id.get, collectionId = tag.id.get))
+          }
+          val badKeeps = applyToKeeps(ownerId, lib, keeps, Set(), saveKeep)
```
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala:179]: code reused!

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala [hunk 8]
```diff
@@ -349,7 +379,7 @@ class LibraryCommander @Inject() (
       case Some(_) =>
         def saveKeep(k: Keep, s: RWSession): Unit = {
           implicit val session = s
-          val newKeep = keepRepo.save(Keep(title = k.title, uriId = k.uriId, url = k.url, urlId = k.urlId,
+          val newKeep = keepRepo.save(Keep(title = k.title, uriId = k.uriId, url = k.url, urlId = k.urlId, isPrivate = k.isPrivate,
```
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala:382]: make sure to keep privacy settings!

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala (hunk 7 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala:179]: Could we `copyKeeps()` instead? As-is, `libraryId` isn't checked for permissions.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala [hunk 9]
```diff
@@ -160,16 +161,29 @@ class LibraryCommander @Inject() (
...
+          Right(badKeeps.toSeq)
+      }
     }
   }
 
-  def getLibraryByUserAndSlug(ownerId: Id[User], slug: LibrarySlug): Option[Library] = {
-    db.readOnlyMaster { implicit s =>
-      libraryRepo.getBySlugAndUserId(userId = ownerId, slug = slug)
-    }
+  def getLibraryByUserAndSlug(ownerId: Id[User], slug: LibrarySlug): Option[Library] = db.readOnlyMaster { implicit s =>
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala:185]: Same for this one, it's just calling a method inside a db session.

It makes it so that if the caller already had a db session open, we're opening two sessions.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala (hunk 8 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala:382]: Good catch.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala (hunk 7 above)
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala:179]: I think in copyKeeps() all tags are copied per keep
Here, we only want to retain the one tag for all new keeps

so if I have one keep with tags "USA" & "Murica"
then I create a library called "USA"
I should have one keep in library "USA" with only one tag of "USA" ? copyKeeps() will also tag new keep with "Murica"

Comment (ahsu1230) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala:179]: as for checking permissions, I should add that.
